foursight_core/react/api/envs.py

```diff
@@ -0,0 +1,82 @@
+import copy
+from typing import Optional
+import os
+from dcicutils import ff_utils
+
+
+class Envs:
+
+    def __init__(self, known_envs: list):
+        # This known_envs should be the list of annotated environment name objects
+        # as returned by app_utils.get_unique_annotated_environment_names, where each
+        # object contains these fields: name, short_name, full_name, public_name, foursight_name
+        self.known_envs = known_envs
+
+    def get_known_envs(self) -> list:
+        return self.known_envs
+
+    def get_known_envs_with_gac_names(self, gac) -> list:
+        known_envs = copy.deepcopy(self.get_known_envs())
+        for known_env in known_envs:
+            known_env["gac_name"] = gac.get_gac_name(known_env["full_name"])
+        return known_envs
+
+    def get_default_env(self) -> str:
+        return os.environ.get("ENV_NAME", "no-default-env")
+
+    def is_known_env(self, env: str) -> bool:
+        return self.find_known_env(env) is not None
+
+    def find_known_env(self, env: str) -> Optional[dict]:
+        if not env:
+            return None
+        env = env.lower()
+        for known_env in self.known_envs:
+            if (known_env["name"].lower() == env
```

Comment:
Envs are case-sensitive. I'm not sure that's good or bad, but I don't advocate tolerating mixed case or uppercase envs because they will not work elsewhere.

Reply:
Hm, yeah, I guess I knew that, will have to undo my case-insensitivity here are elsewhere (FWIW I think this is not good - but that's the way it works elsewhere so will need to make this compatible with that). Will work this out.

Reply:
I partly agree with you, but I think that's a UI issue, not a substrate issue. Common Lisp makes a distinction you might find helpful between 'case-sensitive' and 'case-translating'. That language is case-sensitive, but it is also case-translating, so usually the user is not overwhelmed by the case-sensitivity. I think that's what we want here. Case-sensitive systems are faster because search and comparison don't have to be checking all possible combinations of a lot of cases in every moment, and pointer-comparison is available. So the pipes of our system should be, I think, case-sensitive. I'm totally find with users typing stuff in uppercase if the UI wants to allow that, but the canonicalization can be done one place and does not have to slow down every aspect of the system.

By the way, you can also have the opposite, like the Mac file system, where it is case-preserving but case-insensitive. So it's similar to Linux, except that if you do something like this where you use a filename in the wrong case, something regular Linux would hate:
```
(c4screnv) kentpitman@Kents-MacBook-Pro ~/py/c4-scripts/kmp/utils$ ls -dal [Rr]*
-rw-r--r--  1 kentpitman  staff  3510 Aug 23 17:54 README.rst
(c4screnv) kentpitman@Kents-MacBook-Pro ~/py/c4-scripts/kmp/utils$ head -3 readme.rst
======================
KMP's Utility Commands
======================
```
This is a decision they must have made low-level in the file system, but it means you can 
```
(c4screnv) kentpitman@Kents-MacBook-Pro ~/py/c4-scripts/kmp/utils$  diff readme.rst README.rst
(c4screnv) kentpitman@Kents-MacBook-Pro ~/py/c4-scripts/kmp/utils$ 
```
There's just a performance cost and there's also a need to use completely different functions for comparison, etc. Some languages have a better set of those than others. Python's support for case-insensitive comparison is feeble, whereas by contrast Common Lisp (even though it's often not needed) has really good support for that, since nearly all of the utility operations for search, sort, etc. take an argument of what to use as the item-wise comparison function and there are some built-ins that have twins in case-sensitive and case-insensitive form.